infra/cache/kInfraRedisCacheWrapper.php

```diff
@@ -0,0 +1,314 @@
+	{
+		return $this->callAndDetectErrors('incrby',array($key, $delta));
+	}
+
+	/* (non-PHPdoc)
+	 * @see kBaseCacheWrapper::decrement()
+	 */
+	public function doDecrement($key, $delta = 1)
+	{
+		return $this->callAndDetectErrors('decrby',array($key, $delta));
+	}
+
+	/* (non-PHPdoc)
+    * @see kBaseCacheWrapper::multiGet()
+    */
+	public function doMultiGet($keys)
+	{
+		return $this->callAndDetectErrors('mget',array($keys));
+	}
+
+	public static function sendMonitorEvents()
+	{
+		foreach (self::$_stats as $statsKey => $typeStats)
+		{
+			foreach ($typeStats as $type => $stats)
+			{
+				switch ($type)
+				{
+					case self::STAT_CONN:
+						KalturaMonitorClient::monitorConnTook($statsKey, $stats[self::STAT_TIME], $stats[self::STAT_COUNT]);
+						break;
+
+					case self::STAT_OP:
+						KalturaMonitorClient::monitorRedisAccess($statsKey, $stats[self::STAT_TIME], $stats[self::STAT_COUNT]);
+						break;
+				}
+			}
+		}
+
+		self::$_stats = array();
+	}
+
+	public static function outputStats()
+	{
+		foreach (self::$_stats as $statsKey => $typeStats)
+		{
+			$cur = 'instance:' . $statsKey;
+			foreach ($typeStats as $type => $stats)
+			{
+				foreach ($stats as $key => $value)
+				{
+					$cur .= ', ' . $type . '_' . $key . ':' . $value;
+				}
+			}
+
+			KalturaLog::log($cur);
+		}
+	}
+
+	protected function updateStats($type, $stats)
+	{
+		if (!isset(self::$_stats[$this->statsKey]))
+		{
+			self::$_stats[$this->statsKey] = array();
+		}
+		$typeStats = &self::$_stats[$this->statsKey];
+
+		if (!isset($typeStats[$type]))
+		{
+			$typeStats[$type] = array();
+		}
+
+		foreach ($stats as $key => $value)
+		{
+			if (isset($typeStats[$type][$key]))
+			{
+				$typeStats[$type][$key] += $value;
+			}
+			else
+			{
+				$typeStats[$type][$key] = $value;
+			}
+		}
+	}
+
+	/**
+	 * @param int $errno
+	 * @param string $errstr
+	 * @return bool
+	 */
+	public function errorHandler($errno, $errstr)
+	{
+		self::safeLog("got error from redis [$errno] [$errstr]");
+		$this->gotError = true;
+		return false;
+	}
+
+	protected function callAndDetectErrors($methodName, $params)
+	{
+		while ($this->redis)
+		{
+			$this->gotError = false;
+
+			set_error_handler(array($this, 'errorHandler'));
+			$start = microtime(true);
+			$result = call_user_func_array(array($this->redis, $methodName), $params);
+			$end = microtime(true);
+			restore_error_handler();
+
+			$this->updateStats(self::STAT_OP, array(
+				self::STAT_COUNT => 1,
+				self::STAT_TIME => $end - $start));
+
+			if (!$this->gotError)
+			{
+				return $result;
+			}
+			$this->reconnect();
+		}
+
+		self::safeLog("There is no active redis connection");
+		return false;
+	}
+
+	/**
+	 * @return bool false on error
+	 */
+	protected function reconnect()
+	{
+		$this->close();
+
+		if ($this->connectAttempts >= self::MAX_CONNECT_ATTEMPTS)
+		{
+			return false;
+		}
+
+		$connectResult = false;
+		$connStart = microtime(true);
+		if(!$this->cluster)
```

Comment:
Try and consolidate the code that handles connections to cluster mode and none cluster mode.